File: backend/app/schemas/dispensa.py
"""
Dispensa Schemas
Pydantic models for Dispensa API request/response validation.
"""

from pydantic import BaseModel, Field
from typing import Optional
from uuid import UUID
from datetime import datetime, date


class DispensaItemCreate(BaseModel):
    """Schema for creating a dispensa item"""
    name: str = Field(..., min_length=1, max_length=255, description="Product name")
    quantity: float = Field(1.0, gt=0, description="Quantity")
    unit: Optional[str] = Field(None, max_length=50, description="Unit (pz, kg, g, l, ml)")
    category_id: Optional[UUID] = Field(None, description="Category ID")
    expiry_date: Optional[date] = Field(None, description="Expiry date")
    original_expiry_date: Optional[date] = Field(None, description="Original expiry date before extension")
    barcode: Optional[str] = Field(None, max_length=100, description="Barcode")
    grocy_product_id: Optional[int] = Field(None, description="Grocy product ID")
    grocy_product_name: Optional[str] = Field(None, max_length=255, description="Grocy product name")
    source_item_id: Optional[UUID] = Field(None, description="Source shopping list item ID")
    area_id: Optional[UUID] = Field(None, description="Area ID")
    purchase_price: Optional[float] = Field(None, ge=0, description="Purchase price")
    notes: Optional[str] = Field(None, max_length=500, description="Notes")
    warranty_expiry_date: Optional[date] = Field(None, description="Warranty expiry date")
    trial_expiry_date: Optional[date] = Field(None, description="Trial/return period expiry date")


class DispensaItemUpdate(BaseModel):
    """Schema for updating a dispensa item"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    quantity: Optional[float] = Field(None, gt=0)
    unit: Optional[str] = Field(None, max_length=50)
    category_id: Optional[UUID] = None
    expiry_date: Optional[date] = None
    original_expiry_date: Optional[date] = None
    barcode: Optional[str] = Field(None, max_length=100)
    grocy_product_id: Optional[int] = None
    grocy_product_name: Optional[str] = Field(None, max_length=255)
    area_id: Optional[UUID] = None
    purchase_price: Optional[float] = Field(None, ge=0)
    notes: Optional[str] = Field(None, max_length=500)
    warranty_expiry_date: Optional[date] = None
    trial_expiry_date: Optional[date] = None


class DispensaItemResponse(BaseModel):
    """Schema for dispensa item response"""
    id: UUID
    house_id: UUID
    name: str
    quantity: float
    unit: Optional[str] = None
    category_id: Optional[UUID] = None
    expiry_date: Optional[date] = None
    original_expiry_date: Optional[date] = None
    barcode: Optional[str] = None
    grocy_product_id: Optional[int] = None
    grocy_product_name: Optional[str] = None
    source_list_id: Optional[UUID] = None
    source_item_id: Optional[UUID] = None
    added_by: Optional[UUID] = None
    area_id: Optional[UUID] = None
    purchase_price: Optional[float] = None
    is_consumed: bool
    consumed_at: Optional[datetime] = None
    notes: Optional[str] = None
    warranty_expiry_date: Optional[date] = None
    trial_expiry_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DispensaStatsResponse(BaseModel):
    """Schema for dispensa stats"""
    total: int
    expiring_soon: int
    expired: int


class DispensaItemListResponse(BaseModel):
    """Schema for list of dispensa items"""
    items: list[DispensaItemResponse]
    total: int
    stats: DispensaStatsResponse


class SendToDispensaRequest(BaseModel):
    """Schema for sending shopping list items to dispensa"""
    shopping_list_id: UUID = Field(..., description="Shopping list ID to send items from")
    item_areas: Optional[dict[str, str]] = Field(None, description="Map of item_id → area_id overrides")
    item_expiry_extensions: Optional[dict[str, int]] = Field(None, description="Map of item_id → extension days")


class PreviewFromShoppingListRequest(BaseModel):
    """Schema for previewing items before sending to dispensa"""
    shopping_list_id: UUID = Field(..., description="Shopping list ID to preview")


class PreviewItemResponse(BaseModel):
    """Single item in the preview response"""
    item_id: UUID
    name: str
    quantity: float
    unit: Optional[str] = None
    category_name: Optional[str] = None
    area_id: Optional[UUID] = None
    area_name: Optional[str] = None


class PreviewAreaResponse(BaseModel):
    """Area info in the preview response"""
    id: UUID
    name: str
    icon: Optional[str] = None
    expiry_extension_enabled: bool = False
    disable_expiry_tracking: bool = False
    warranty_tracking_enabled: bool = False
    default_warranty_months: Optional[int] = None
    trial_period_enabled: bool = False
    default_trial_days: Optional[int] = None


class PreviewFromShoppingListResponse(BaseModel):
    """Response for preview-from-shopping-list"""
    items: list[PreviewItemResponse]
    areas: list[PreviewAreaResponse]


class ConsumeItemRequest(BaseModel):
    """Schema for consuming an item (partial or total)"""
    quantity: Optional[float] = Field(None, gt=0, description="Quantity to consume (None = consume all)")

File: frontend/src/services/dispensa.ts
import api from './api'
import type { DispensaItem, DispensaStats } from '@/types'

interface DispensaItemListResponse {
  items: DispensaItem[]
  total: number
  stats: DispensaStats
}

interface SendToDispensaResponse {
  message: string
  count: number
}

export interface PreviewItem {
  item_id: string
  name: string
  quantity: number
  unit: string | null
  category_name: string | null
  area_id: string | null
  area_name: string | null
}

export interface PreviewArea {
  id: string
  name: string
  icon: string | null
  expiry_extension_enabled: boolean
  disable_expiry_tracking: boolean
  warranty_tracking_enabled: boolean
  default_warranty_months: number | null
  trial_period_enabled: boolean
  default_trial_days: number | null
}

export interface PreviewFromShoppingListResponse {
  items: PreviewItem[]
  areas: PreviewArea[]
}

interface GetItemsParams {
  search?: string
  category_id?: string
  expiring?: boolean
  expired?: boolean
  consumed?: boolean
  show_all?: boolean
  area_id?: string
}

export const dispensaService = {
  /**
   * Get all dispensa items with optional filters
   */
  async getItems(houseId: string, params?: GetItemsParams): Promise<DispensaItemListResponse> {
    const response = await api.get('/dispensa', {
      params: { house_id: houseId, ...params },
    })
    return response.data
  },

  /**
   * Get dispensa statistics
   */
  async getStats(houseId: string): Promise<DispensaStats> {
    const response = await api.get('/dispensa/stats', {
      params: { house_id: houseId },
    })
    return response.data
  },

  /**
   * Create a new dispensa item
   */
  async createItem(houseId: string, data: {
    name: string
    quantity?: number
    unit?: string
    category_id?: string
    expiry_date?: string
    original_expiry_date?: string
    barcode?: string
    grocy_product_id?: number
    grocy_product_name?: string
    area_id?: string
    purchase_price?: number
    notes?: string
    warranty_expiry_date?: string
    trial_expiry_date?: string
  }): Promise<DispensaItem> {
    const response = await api.post('/dispensa', data, {
      params: { house_id: houseId },
    })
    return response.data
  },

  /**
   * Update a dispensa item
   */
  async updateItem(houseId: string, itemId: string, data: {
    name?: string
    quantity?: number
    unit?: string
    category_id?: string
    expiry_date?: string
    barcode?: string
    notes?: string
    area_id?: string
  }): Promise<DispensaItem> {
    const response = await api.put(`/dispensa/${itemId}`, data, {
      params: { house_id: houseId },
    })
    return response.data
  },

  /**
   * Delete a dispensa item
   */
  async deleteItem(houseId: string, itemId: string): Promise<void> {
    await api.delete(`/dispensa/${itemId}`, {
      params: { house_id: houseId },
    })
  },

  /**
   * Consume an item (total or partial)
   */
  async consumeItem(houseId: string, itemId: string, quantity?: number): Promise<DispensaItem> {
    const response = await api.post(`/dispensa/${itemId}/consume`,
      { quantity: quantity ?? null },
      { params: { house_id: houseId } },
    )
    return response.data
  },

  /**
   * Restore a consumed item
   */
  async unconsumeItem(houseId: string, itemId: string): Promise<DispensaItem> {
    const response = await api.post(`/dispensa/${itemId}/unconsume`, null, {
      params: { house_id: houseId },
    })
    return response.data
  },

  /**
   * Preview items from a shopping list with resolved areas
   */
  async previewFromShoppingList(houseId: string, listId: string): Promise<PreviewFromShoppingListResponse> {
    const response = await api.post('/dispensa/preview-from-shopping-list',
      { shopping_list_id: listId },
      { params: { house_id: houseId } },
    )
    return response.data
  },

  /**
   * Send verified items from a shopping list to the dispensa
   */
  async sendFromShoppingList(houseId: string, listId: string, itemAreas?: Record<string, string>, itemExpiryExtensions?: Record<string, number>): Promise<SendToDispensaResponse> {
    const response = await api.post('/dispensa/from-shopping-list',
      {
        shopping_list_id: listId,
        item_areas: itemAreas || null,
        item_expiry_extensions: itemExpiryExtensions || null,
      },
      { params: { house_id: houseId } },
    )
    return response.data
  },
}

export default dispensaService
